### ISSUE TITLE
tf.nn.sigmoid loses a finite first-derivative signal for float64 input

### ISSUE THREAD

**Issue by ALinrunrun** [tensorflow/tensorflow#126631]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

Python 3.13

### Current behavior?

`tf.nn.sigmoid` returns zero for a first-derivative component whose correctly rounded `float64` reference value is finite, normal, and nonzero.

For sigmoid, the derivative is `sigmoid(x) * (1 - sigmoid(x))`, or equivalently a stable form such as `exp(-x) / (1 + exp(-x))**2` for positive `x`. At `x = 37.42994775023705`, the high-precision reference rounded to float64 is approximately `5.551115123125775e-17`.

However, TensorFlow autodiff returns `0.0` for this derivative. The forward value rounds to `1.0`, but the derivative remains representable in float64 and should not be lost during the first-order autodiff pass.

Expected behavior?

The first derivative of `tf.nn.sigmoid` at `x = 37.42994775023705` should be close to `5.551115123125775e-17`, not `0.0`.

### Standalone code to reproduce the issue

```shell
import os
os.environ.setdefault("CUDA_VISIBLE_DEVICES", "-1")
os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "3")

import tensorflow as tf

x_value = 37.42994775023705
expected = 5.551115123125775e-17
rtol = 1e-6

x = tf.Variable(x_value, dtype=tf.float64)

with tf.GradientTape() as tape:
    y = tf.nn.sigmoid(x)

g = tape.gradient(y, x)

actual = float(g)
rel_err = abs(actual - expected) / abs(expected)

print(f"x: {x_value!r}")
print(f"forward value: {float(y)!r}")
print(f"tensorflow gradient: {actual!r}")
print(f"expected gradient: {expected!r}")
print(f"relative error: {rel_err:.3e}")

if rel_err > rtol:
    print("BUG REPRODUCED: tf.nn.sigmoid loses a finite first-derivative signal")
else:
    print("not reproduced")
```

### Relevant log output

```shell
x: 37.42994775023705
forward value: 1.0
tensorflow gradient: 0.0
expected gradient: 5.551115123125775e-17
relative error: 1.000e+00
BUG REPRODUCED: tf.nn.sigmoid loses a finite first-derivative signal
```

**Comment by Venkat6871** [comment]
Hi @ALinrunrun,

Thank you for reporting this issue and for providing a clear, reproducible example. You've demonstrated that `tf.nn.sigmoid`'s first derivative returns `0.0` for a `float64` input where a small, finite value is expected, indicating a potential loss of precision in the gradient calculation for this edge case.

We appreciate the detailed explanation and reproduction. We will investigate this issue from our side and provide an update.

**Comment by Venkat6871** [comment]
Hi @ALinrunrun  ,
Thank you for raising your concern here.
A fix for this issue has been proposed in this [PR](https://github.com/tensorflow/tensorflow/pull/127010) Once it is merged, this issue will be resolved.

Hi @bizstpa ,
Thank you for raising the PR to address this issue. We truly appreciate your initiative and the effort in contributing the fix.
